src/ModelManager.ts

```diff
@@ -125,96 +147,181 @@ export class ModelManager {
+        this._fetchPromises = {};
+        this._initPromise = null;
+
+        this._modelClient = ((config && config.modelClient) || new ModelClient());
 
-        this._clientlibUtil = new AuthoringUtils(this.modelClient.apiHost);
         this._editorClient = new EditorClient(this);
-        this._modelStore = (initialModel) ? new ModelStore(rootModelPath, initialModel) : new ModelStore(rootModelPath);
+        this._clientlibUtil = new AuthoringUtils(this.modelClient.apiHost);
+    }
 
-        this._initPromise = this._checkDependencies().then(() => {
-            const data = this.modelStore.getData(rootModelPath);
+    /**
+     * Returns paths required for fetching root model
+     */
+    private _getPathsForModel(config?: ModelManagerConfiguration) {
+        // Model path explicitly provided by user in config
+        const path = config && config.path;
 
-            if (data && (Object.keys(data).length > 0)) {
-                triggerPageModelLoaded(data);
+        // Model path set statically via meta property
+        const pageModelRoot = PathUtils.getMetaPropertyValue(MetaProperty.PAGE_MODEL_ROOT_URL);
+        const metaPropertyModelUrl = PathUtils.internalize(pageModelRoot);
 
-                return data;
-            } else {
-                return this._fetchData(rootModelURL).then((rootModel: Model) => {
-                    this.modelStore.initialize(rootModelPath, rootModel);
+        const aemApiHost = this.modelClient.apiHost;
+        const isRemoteApp = PathUtils.isBrowser() && aemApiHost && (PathUtils.getCurrentURL() !== aemApiHost);
+        const currentPathname = !isRemoteApp ? PathUtils.getCurrentPathname() : '';
+        // For remote apps in edit mode, to fetch path via parent URL
 
-                    // Append the child page if the page model doesn't correspond to the URL of the root model
-                    // and if the model root path doesn't already contain the child model (asynchronous page load)
-                    if (!!currentPathname && !!sanitizedCurrentPathname) {
-                        if (!isPageURLRoot(currentPathname, metaPropertyModelUrl) && !hasChildOfPath(rootModel, currentPathname)) {
-                            return this._fetchData(currentPathname).then((model: any) => {
-                                this.modelStore.insertData(sanitizedCurrentPathname, model);
+        const sanitizedCurrentPathname = ((currentPathname && PathUtils.sanitize(currentPathname)) || '') as string;
 
-                                const data = this.modelStore.getData();
+        // Fetch the app root model
+        // 1. consider the provided page path
+        // 2. consider the meta property value
+        // 3. fallback to the model path contained in the URL for the default SPA
+        const rootModelURL = path || metaPropertyModelUrl || sanitizedCurrentPathname;
+        const rootModelPath = PathUtils.sanitize(rootModelURL) || '';
+
+        this._modelPaths = {
```

Comment:
Is this operation performed in `_getPathsForModel`? Also, it seems that no object is returned from this getter. Am I misreading?

Reply:
Yes, but it would be cleaner if done within _initializeFields. I'll move this.